### PR TITLE
AE-419: Agent that checks age of AAVE price oracle against maximum threshold

### PR DESCRIPTION
**Pull request by echoAE**
This agent examines the age of the AAVE price oracles (Chainlink aggregators) to determine if they exceed the maximum threshold (24 hours = 86400 seconds).  For each AAVE lending pool reserve token, if the age of the oracle exceeds that threshold, an alert is generated.

### CODE REVIEW

File: interfaces/IPriceOracle.json [hunk 1]
```diff
@@ -3,42 +3,242 @@
...
+			"inputs": [
+				{
+					"internalType": "address",
+					"name": "newOwner",
+					"type": "address"
+				}
+			],
+			"name": "transferOwnership",
+			"outputs": [],
+			"stateMutability": "nonpayable",
```
Comment (y0wl) [inline interfaces/IPriceOracle.json:240]: This should not be put inside this json file, there is a standalone that accurately represents the abi

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  LendingPoolAddressesProvider: lendingPoolAddressesProvider,
+  ProtocolDataProvider: protocolDataProviderAddress,
+} = require('../../contract-addresses.json');
+const { abi: protocolDataProviderAbi } = require('../../interfaces/AaveProtocolDataProvider.json');
+const { abi: priceOracleAbi } = require('../../interfaces/IPriceOracle.json');
+const { abi: chainlinkAggregatorAbi } = require('../../interfaces/IChainlinkAggregator.json');
+const { abi: lendingPoolAddressesProviderAbi } = require('../../interfaces/ILendingPoolAddressesProvider.json');
+
+// set up the an ethers provider
+const jsonRpcProvider = new ethers.providers.WebSocketProvider(getJsonRpcUrl());
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.js:18]: this needs to change from `WebSocketProvider` to `getDefaultProvider`

**Review by arbitrary-cr0 — CHANGES_REQUESTED** [review]

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+// time threshold over which we trigger alerts (24 hours = 86400 seconds)
+// this value comes from the Chainlink web interface for price feeds (mouseover Trigger parameters)
+//  'A new trusted answer is written when the off-chain price moves more than the deviation
+//   threshold or 86400 seconds have passed since the last answer was written on-chain.'
+const ORACLE_AGE_THRESHOLD_SECONDS = 86400;
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.js:24]: should go into `agent-config.json`

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+      // the updatedAt value is of type ethers.BigNumber
+      // ethers.BigNumber is not the same as BigNumber from bignumber.js
+      // therefore, we need to convert from ethers.BigNumber to BigNumber
+      const timestamp = new BigNumber(roundData.updatedAt.toString());
+
+      // calculate the difference between the current block timestamp and the last oracle update
+      const oracleAge = blockTimestamp.minus(timestamp);
+
+      if (oracleAge.isGreaterThan(ORACLE_AGE_THRESHOLD_SECONDS)) {
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.js:138]: I personally am a fan of globals being all caps, should we start doing this and maybe refactor old code to match this?

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+  return tokenAddressContractTuples;
+}
+
+function provideHandleBlock(tokensAddressesContractsPromise) {
+  return async function handleBlock(blockEvent) {
+    const findings = [];
+
+    // settle the promise the first time, all subsequent times just get the resolve() value
+    const tokensAddressesContracts = await tokensAddressesContractsPromise;
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.js:110]: no error checking on this, do we care? I think no :) up to you

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        severity: FindingSeverity.Medium,
+        type: FindingType.Suspicious,
+        everestId: '0xa3d1fd85c0b62fa8bab6b818ffc96b5ec57602b6',
+        metadata: {
+          symbol: token.symbol,
+          tokenAddress: token.tokenAddress,
+          oracleAge: oracleAgeTooOld,
+          priceSourceAddress,
+        },
+      });
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js:95]: Something I like doing is moving this into a shared location or even importing it from the agent itself. Keeps code dupe down and makes it so you only have to update 1 spot if we update this in the future /shrug

**Review by y0wl — CHANGES_REQUESTED** [review]

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Done.

File: .eslintrc.js [hunk 7]
```diff
@@ -11,6 +11,7 @@ module.exports = {
   parserOptions: {
     ecmaVersion: 12,
   },
+	ignorePatterns: ['.eslintrc.js'],
```
[the remark this answers is not in the document]
Comment (echoAE) [reply .eslintrc.js]: Removed.  I added the fixes that eslint made to the `.eslintrc.js` file.

File: interfaces/IPriceOracle.json (hunk 1 above)
Comment (echoAE) [reply interfaces/IPriceOracle.json]: Removed from `IPriceOracle.json` and `AaveOracle.json` added.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 2 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Done.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 3 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Moved.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 5 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js:116]: I agree that if this fails, we should just let it fail.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 4 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js:144]: I'm importing this variable from `agent-config.json` where it's lowerCamelCase and I'm importing it as all caps.  I support all caps signifying globals.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js (hunk 6 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js]: I've removed the duplicated code and added an import of `createAlert()` in the `.spec.js` file.

**Review by arbitrary-cr0 — APPROVED** [review]

File: agent-config.json [hunk 8]
```diff
@@ -12,5 +12,8 @@
     "numStds" : 3,
     "minElements": 40
   },
+	"check-reserve-update-frequency" : {
+		"oracleAgeThresholdSeconds": 86400
+	},
```
Comment (y0wl) [inline agent-config.json]: spaces not tabs por favor
```vim
autocmd FileType javascript setlocal shiftwidth=2 tabstop=2
autocmd FileType json setlocal shiftwidth=2 tabstop=2```

adding this to your vimrc will make life eaassy

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 9]
```diff
@@ -10,18 +10,24 @@ const {
...
-// set up the an ethers provider
-const jsonRpcProvider = new ethers.providers.WebSocketProvider(getJsonRpcUrl());
-
 // time threshold over which we trigger alerts (24 hours = 86400 seconds)
 // this value comes from the Chainlink web interface for price feeds (mouseover Trigger parameters)
 //  'A new trusted answer is written when the off-chain price moves more than the deviation
 //   threshold or 86400 seconds have passed since the last answer was written on-chain.'
-const ORACLE_AGE_THRESHOLD_SECONDS = 86400;
+const {
+  oracleAgeThresholdSeconds: ORACLE_AGE_THRESHOLD_SECONDS,
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.js]: this is looking for a key inside another key, the agent-config does not expose the `oracleAgeThresholdSeconds` key directly

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js [hunk 10]
```diff
@@ -1,12 +1,11 @@
...
 const {
-  Finding,
-  FindingSeverity,
-  FindingType,
-  createBlockEvent,
-} = require('forta-agent');
-const { provideHandleBlock, teardownProvider } = require('./check-reserve-update-frequency');
+  provideHandleBlock,
+  teardownProvider,
+  createAlert,
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js:7]: you import `createAlert` but your agent never exports this in its `module.exports`

File: agent-config.json (hunk 8 above)
Comment (echoAE) [reply agent-config.json]: `expandtab` added for `json`, `javascript`, and `typescript`

More importantly, tabs replaced with spaces.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 9 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Import fixed.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js (hunk 10 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js:7]: `createAlert` added to exports.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+} = require('./check-reserve-update-frequency');
+
+describe('AAVE reserve price oracle agent', () => {
+  let handleBlock;
+
+  afterEach(() => {
+    teardownProvider();
+  });
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js:15]: change this to `afterAll`

**Review by y0wl — CHANGES_REQUESTED** [review]

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+];
+
+// helper function to create alerts
+function createAlert(reserveToken, oracleAge, priceSourceAddress) {
+  return Finding.fromObject({
+    name: `Stale AAVE Price Oracle Data for ${reserveToken.symbol}`,
+    description: `Token ${reserveToken.symbol} Price Oracle Age: ${oracleAge} seconds`,
+    alertId: 'AE-AAVE-PRICE-ORACLE-STALE',
+    severity: FindingSeverity.Medium,
+    type: FindingType.Suspicious,
```
Comment (arbitrary-kdawg) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.js:52]: Would FindingType.Degraded be more suitable in this case?

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+      // create the mocked promise to pass in to provideHandleBlock
+      const promise = await mockTokensAddressesContractPromise(updatedAtTimestamp);
+
+      // create the block handler
+      handleBlock = provideHandleBlock(promise);
+
+      // create expected finding
+      const token = promise[0][0];
+      const priceSourceAddress = promise[0][1];
```
Comment (y0wl) [inline src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js:79]: this is incredibly confusing :( I understand it but some suggestions:

 - we don't have tests for more than 1 reserve, maybe the first list can be removed?
 - using keyword assignments in an object would make this so much easier to read
on line 47 change `return [token, address, contract];` to `return {token, address, contract};`
and here you can use `const { token, priceSourceAddress } = promise;`
 - I'm also not a huge fan of the var name `promise` but meh

**Review by y0wl — CHANGES_REQUESTED** [review]

**Comment by y0wl** [comment]
this has work that needs to reflect changes in #11, my suggestion is remove the eslint overrides and make the few changes to the provider creation and teardown, it's teeny tiny :)

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js (hunk 12 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Good suggestion.  I switched it.

File: interfaces/IChainlinkAggregator.json [hunk 14]
```diff
@@ -3,128 +3,514 @@
   "contractName": "IChainlinkAggregator",
```
[the remark this answers is not in the document]
Comment (echoAE) [reply interfaces/IChainlinkAggregator.json]: I reverted my incorrect changes to the IChainlinkAggregator abi and added the AggregatorV3Interface abi from here:
https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol

Also, no tabs.

File: interfaces/AaveOracle.json [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+      "name": "transferOwnership",
+      "outputs": [],
+      "stateMutability": "nonpayable",
+      "type": "function"
+    }
+  ],
+	"bytecode": "",
+	"deployedBytecode": "",
+	"linkReferences": {},
+	"deployedLinkReferences": {}
```
[the remark this answers is not in the document]
Comment (echoAE) [reply interfaces/AaveOracle.json]: Tabs replaced with spaces.

File: interfaces/AaveOracle.json [hunk 16]
```diff
@@ -0,0 +1,248 @@
+{
+	"_format": "hh-sol-artifact-1",
+	"contractName": "AaveOracle",
+	"sourceName": "contracts/misc/AaveOracle.sol",
```
[the remark this answers is not in the document]
Comment (echoAE) [reply interfaces/AaveOracle.json]: Tabs replaced with spaces.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+} = require('../../contract-addresses.json');
+const { abi: protocolDataProviderAbi } = require('../../interfaces/AaveProtocolDataProvider.json');
+const { abi: aaveOracleAbi } = require('../../interfaces/AaveOracle.json');
+const { abi: chainlinkAggregatorAbi } = require('../../interfaces/IChainlinkAggregator.json');
+const { abi: lendingPoolAddressesProviderAbi } = require('../../interfaces/ILendingPoolAddressesProvider.json');
+
+// time threshold over which we trigger alerts (24 hours = 86400 seconds)
+// this value comes from the Chainlink web interface for price feeds (mouseover Trigger parameters)
+//  'A new trusted answer is written when the off-chain price moves more than the deviation
+//   threshold or 86400 seconds have passed since the last answer was written on-chain.'
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Comment moved to just above the variable assignment.  Note added to indicate that the value originates from the `agent-config.json` file.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+  aaveEverestId: AAVE_EVEREST_ID,
+} = require('../../agent-config.json');
+
+const ORACLE_AGE_THRESHOLD_SECONDS = checkReserveUpdateFrequency.oracleAgeThresholdSeconds;
+
+// set up the an ethers provider
+// use ethers.providers.getDefaultProvider() in lieu of ethers.providers.WebSocketProvider()
+// websockets are not supported in production
+// eslint-disable-next-line new-cap
+const jsonRpcProvider = new ethers.providers.getDefaultProvider(getJsonRpcUrl());
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Changed from getDefaultProvider() to JsonRpcProvider().

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+
+    return findings;
+  };
+}
+
+// closes the ethers provider websocket
+async function teardownProvider() {
+  await jsonRpcProvider.destroy();
+}
+
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Removed teardownProvider() from code.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.js [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+module.exports = {
+  provideHandleBlock,
+  handleBlock: provideHandleBlock(initializeTokensContracts()),
+  teardownProvider,
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.js]: Removed teardownProvider() from code.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js [hunk 21]
```diff
@@ -0,0 +1,110 @@
+// required libraries
+const BigNumber = require('bignumber.js');
+const { createBlockEvent } = require('forta-agent');
+const {
+  provideHandleBlock,
+  teardownProvider,
```
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js]: Removed teardownProvider() from code.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js]: Removed teardownProvider() from code.

File: src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js (hunk 13 above)
Comment (echoAE) [reply src/check-reserve-update-frequency/check-reserve-update-frequency.spec.js]: Replaced nested Array with key/value pairs, as suggested.
Renamed `promise` to `tokensAddressesContracts` to indicate it is an Array of those values.

I'm not completely certain on the comment stating `maybe the first list can be removed`, but I'm happy to consider the suggestion after clarification.

**Comment by echoAE** [comment]
eslint overrides (`ignorePatterns`) have been removed and the provider creation/teardown changes have been made.

**Review by arbitrary-kdawg — APPROVED** [review]

**Review by y0wl — APPROVED** [review]

**Pull request merged**